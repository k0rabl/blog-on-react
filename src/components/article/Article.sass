.article
  margin-top: 25px
  margin-bottom: 25px
  .top
    margin-bottom: 15px
    display: flex
    justify-content: space-between
    .title
      font-size: 26px