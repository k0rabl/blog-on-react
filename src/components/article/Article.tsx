import React, { Component } from 'react'
import IArticle from '../../interfaces/IArticle'

class Article extends Component<IArticle> {
  constructor(props: IArticle){
    super(props)
  }

  render() {
    const { date, name, desc } = this.props
    
    return (
      <div className="container">
        <div className="top">
          <div className="title">{name}</div>
          <div className="date">{date.toDateString()}</div>
        </div>
        <div className="bottom">
          <div className="image"></div>
          <div className="descr">{desc}</div>
        </div>
      </div>
    )
  }
}

export default Article