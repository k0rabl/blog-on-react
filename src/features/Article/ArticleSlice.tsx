import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import Articles from '../../fixtures/Articles'
import IArticle from './component/IArticle'

export interface PaginationState {
  filteredArticles: IArticle[]
  articles: IArticle[]
}

const initialState: PaginationState = {
  filteredArticles: Articles,
  articles: Articles
}

export const articlesSlice = createSlice({
  name: 'search',
  initialState,
  reducers: {
    handleRead: (state, action: PayloadAction<number>) => {
      state.articles =  state.articles.map((element: IArticle) => ({
        ...element,
        isRead: element.id === action.payload ? true : element.isRead
      }))
    },
    handleDeleteElement: (state, action: PayloadAction<number>) => {
      state.articles = state.articles.filter(element => {
        return element.id !== action.payload
      })
    },
    handleEditElement: (state, action: PayloadAction<IArticle>) => {
      state.articles = state.articles.map((element: IArticle) => {
        if(element.id === action.payload.id)  
          return action.payload

        return element
      })
    },
    handleAddElement: (state, action: PayloadAction<IArticle>) => {
      state.articles.push(action.payload)
    }
  }
})

export const { 
  handleRead, 
  handleDeleteElement,
  handleEditElement,
  handleAddElement
} = articlesSlice.actions
export default articlesSlice.reducer