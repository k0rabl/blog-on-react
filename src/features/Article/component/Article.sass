
.article
  margin-top: 25px
  margin-bottom: 25px
  width: 100%
  display: flex
  justify-content: space-between
  .leftImage
    width: 30%
    border: 1px solid #b10303
    margin-right: 15px
    display: flex
    justify-content: center
    align-items: center
    img
      width: 100%
      height: 200px
  .rightDescr
    width: 70%
  &Link
    color: inherit
  
  .top
    margin-bottom: 15px
    display: flex
    justify-content: space-between
    &-left
      display: flex
      align-items: center
      .title
        margin-right: 15px
        font-size: 26px
        color: #b10303
        transition: color .2s
        &.edit
          color: #B94A3E
          white-space: nowrap
          text-overflow: ellipsis
          max-width: 350px
          overflow: hidden
        &.read
          color: #B94A3E
        &:hover
          color: #742c24
      .btn
        &-delete
          margin-right: 15px