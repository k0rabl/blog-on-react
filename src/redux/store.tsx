import { configureStore } from "@reduxjs/toolkit"

import paginationSlice from "../features/Pagination/PaginationSlice"
import searchSlice from "../features/Article/ArticleSlice"
import modeSlice from "../features/edit/modeSlice"

export const store = configureStore({
  reducer: {
    pagination: paginationSlice,
    search: searchSlice,
    edit: modeSlice
  }
})

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch
