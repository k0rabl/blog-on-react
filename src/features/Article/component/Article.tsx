import React, { Component } from 'react'
import IArticle from './IArticle'

import './Article.sass'
import { Link } from 'react-router-dom'

class Article extends Component<IArticle> {

  render() {
    const { id, date, name, desc, isRead } = this.props
    
    return (
      <Link to={`/${id}`} className="link articleLink">
        <div className="container article">
          <div className="top">
            <div className={`title ${isRead && 'read'}`}>{name}</div>
            <div className="date">{date}</div>
          </div>
          <div className="bottom">
            <div className="image"></div>
            <div className="descr">{desc}</div>
          </div>
        </div>
      </Link>
    )
  }
}

export default Article