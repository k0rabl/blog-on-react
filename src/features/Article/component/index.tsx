import React, { PureComponent } from 'react'
import { ArticleProps, IState } from './IArticle'

import './Article.sass'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import { RootState } from '../../../redux/store'
import { handleDeleteElement } from '../ArticleSlice'
import { Modal } from '../../Modal'
import AlertContext from '../../../AlertContext'

class Article extends PureComponent<ArticleProps, IState> {
  static contextType = AlertContext

  constructor(props: ArticleProps){
    super(props)

    this.state = {
      opened: false
    }
  }

  handleDelete = () => {
    const { setOpen, setType } = this.context 
    const { id, handleDeleteElement } = this.props
    
    handleDeleteElement(Number(id))
    
    setOpen(true)
    setType('Delete')
    this.setState({opened: false})
  }



  handleShowModal = () => {
    const { opened } = this.state
    this.setState({opened: !opened})
  }

  render() {
    const { 
      id, 
      date, 
      name, 
      desc, 
      isRead,
      editMode
    } = this.props   
    
    return (
      <div className="container article">
        <div className="top">
          <div className="top-left">
            
            <Link to={`/${id}`} className="link articleLink">
              <div className={`title ${isRead && 'read'} ${editMode && 'edit'}`}>{name}</div>
            </Link>
            
            {editMode && 
              <div className="buttons">
                <button  className="waves-effect waves-light redC-bg lighten-3 btn btn-delete" onClick={this.handleShowModal}>
                  <i className="material-icons center">delete</i>
                </button > 
                
                {/* <button  className="waves-effect waves-light redC-bg lighten-3 btn btn-edit" onClick={() => console.log('edit')}>
                  <i className="material-icons center">edit</i>
                </button > */} 
              </div>
            }
          </div>
          
          <div className="date">{date}</div>
        </div>
        <div className="bottom">
          <div className="image"></div>
          <div className="descr">{desc}</div>
        </div>
        <Modal 
          opened={this.state.opened}
          onClose={this.handleShowModal} 
          onAccept={this.handleDelete} 
          type='delete'
        />
      </div>
    )
  }
}


const mapStateToProps = (state: RootState) => ({
  editMode: state.edit.editMode,
  articles: state.search.filteredArticles
})

const mapDispatch = { handleDeleteElement }

export default connect(mapStateToProps, mapDispatch)(Article)