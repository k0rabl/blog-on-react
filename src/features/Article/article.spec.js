import React from "react"
import Article from  './component/index'
import { shallow } from 'enzyme';

import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { store } from '../../redux/store'
import { Provider } from 'react-redux'

Enzyme.configure({ adapter: new Adapter() });

const setUp = props => shallow(
  <Provider store={store}>
    <Article {...props} />
  </Provider>
)

describe('<Article />', () => {
  let component;
  beforeEach(() => {
    component = setUp()
  })

  it('renders <Article /> component', () => {
    expect(component).toMatchSnapshot()
  })
})