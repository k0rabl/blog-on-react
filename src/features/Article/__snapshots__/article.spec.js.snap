// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Article /> renders <Article /> component 1`] = `ShallowWrapper {}`;
